### PR TITLE
[flutter_tools] Add namespace getter in Android project; use namespace as fallback

### LINKED ISSUE #121404: [flutter_tools] Support `namespace` attribute introduced in AGP 7.3
From AGP(android gradle plugin) 7.3 onwards it is recommended to replace the use of the `package` attribute in the `Manifest.xml` with the `namespace` attribute in the build.gradle file.

We should update the tooling to support this new attribute.

1) Check the `build.gradle` for a `namespace` attribute
2) If it is not found, fall back to checking the Manifest's `package` attribute

https://developer.android.com/guide/topics/manifest/manifest-element (see the last phrase under `package`)
https://developer.android.com/studio/build/configure-app-module#set-namespace

## Steps to Reproduce

1. `flutter create myapp`
2. Remove the `package` attribute from the Android Manifest
3. In `android/app/build.gradle` use the new `namespace` attribute:

```
android {
  namespace '<value of the original package attribute from the Manifest>'
}
```
4. `flutter run`
5. The app fails to build (see the logs below)

## Logs

<details>
<summary>Logs</summary>

```
navaronbracke@MacBook-Pro-van-Navaron weforza % flutter run
Launching lib/main.dart on motorola one in debug mode...
package identifier or launch activity not found.
Please check /Users/navaronbracke/Documents/weforza/android/app/src/main/AndroidManifest.xml for errors.
No application found for TargetPlatform.android_arm64.
Is your project missing an android/app/src/main/AndroidManifest.xml?
Consider running "flutter create ." to create one.
```

```
[✓] Flutter (Channel stable, 3.7.3, on macOS 13.2 22D49 darwin-x64, locale en-BE)
    • Flutter version 3.7.3 on channel stable at /Users/navaronbracke/Documents/flutter
    • Upstream repository git@github.com:navaronbracke/flutter.git
    • FLUTTER_GIT_URL = git@github.com:navaronbracke/flutter.git
    • Framework revision 9944297138 (2 weeks ago), 2023-02-08 15:46:04 -0800
    • Engine revision 248290d6d5
    • Dart version 2.19.2
    • DevTools version 2.20.1

[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.0)
    • Android SDK at /Users/navaronbracke/Library/Android/sdk
    • Platform android-33, build-tools 33.0.0
    • ANDROID_HOME = /Users/navaronbracke/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.13+0-b1751.21-8125866)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 14.2)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 14C18
    • CocoaPods version 1.11.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2021.3)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.13+0-b1751.21-8125866)

[✓] VS Code (version 1.75.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.58.0

[✓] Connected device (2 available)
    • macOS (desktop) • macos  • darwin-x64     • macOS 13.2 22D49 darwin-x64
    • Chrome (web)    • chrome • web-javascript • Google Chrome 110.0.5481.177

[✓] HTTP Host Availability
    • All required HTTP hosts are available

• No issues found!
```

</details>

### PR DESCRIPTION
**Pull request by navaronbracke**
For newer Android projects the `package` attribute in the Manifest has been replaced with a `namespace` attribute under the `android` section of the build.gradle file.

*List which issues are fixed by this PR. You must list at least one issue.*
Fixes: #121404

*If you had to change anything in the [flutter/tests] repo, include a link to the migration guide as per the [breaking change policy].*
N/A

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by navaronbracke** [comment]
@christopherfujino Could you give me a hint for a test? I'd want to have a stub `build.gradle` with the namespace tag in it and a manifest without a `package` attribute. Then I'd try to build an apk and assert on the package identifier?

I just don't know how to do that exactly.

File: packages/flutter_tools/test/general.shard/android/android_gradle_builder_test.dart [hunk 1]
```diff
@@ -716,6 +716,96 @@ void main() {
...
+          ),
+        ),
+        target: 'lib/main.dart',
+        isBuildingBundle: false,
+        localGradleErrors: const <GradleHandledError>[],
+      );
+
+      expect(processManager, hasNoRemainingExpectations);
+
+      // TODO: verify the packageId of the built APK equals the namespace value
```
Comment (navaronbracke) [inline packages/flutter_tools/test/general.shard/android/android_gradle_builder_test.dart:804]: So this test verifies that building an APK using the namespace works, but I'd like to assert that the namespace value is also correct. Any tips on how to do that? Or is that not required (as the test already validates that the APK is built without errors)

Comment (christopherfujino) [reply packages/flutter_tools/test/general.shard/android/android_gradle_builder_test.dart:804]: I don't think this test validates your change--I reverted your lib/src changes and this test still passes.

Can you test `AndroidApk.fromAndroidProject()`?

File: packages/flutter_tools/lib/src/project.dart [hunk 2]
```diff
@@ -486,9 +487,15 @@ class AndroidProject extends FlutterProjectPlatform {
   }
 
   File get appManifestFile {
-    return isUsingGradle
```
Comment (navaronbracke) [inline packages/flutter_tools/lib/src/project.dart:489]: @christopherfujino I had mismatching results when originally running this test.
The `fileSystem` created in the test would report that the `appManifestFile` exists, but the `globals.fs` file system in this getter reported that it did not exist.

According to a doc comment I can override `globals.fs` in tests. I don't see the `setUp((){})` doing something like that. What would be the right thing to do here?

A) Override `globals.fs` in the test
B) Fix the getter to use `hostAppGradleRoot`, which does use the `fileSystem` that was set up in the specific test (which is what I did here)

File: packages/flutter_tools/test/general.shard/android/android_gradle_builder_test.dart [hunk 3]
```diff
@@ -716,6 +723,59 @@ void main() {
       AndroidStudio: () => FakeAndroidStudio(),
     });
 
+    testUsingContext('Uses namespace attribute if manifest lacks a package attribute', () async {
```
Comment (navaronbracke) [inline packages/flutter_tools/test/general.shard/android/android_gradle_builder_test.dart:726]: @christopherfujino I'm having some weird results with running this test in VSCode, do you have an idea why this is happening?

Disclaimer: I ran only this test in all cases listed below, so it should not have anything to do with test order and/or sharding.

a) If I put `await AndroidApk.fromProject();` in comment and add an `expect(true, true)` at the end of the test, it passes as expected.
b) If I put `await AndroidApk.fromProject();` in comment and add an `expect(true, false)` at the end, it hangs indefinitely (it should have failed instantly)
c) If I keep `await AndroidApk.fromProject();` but don't have an `expect()` it passes instantly.
d) If I keep `await AndroidApk.fromProject();` and have an `expect(apk?.id, 'com.example.foo')`, it hangs indefinitely.

File: packages/flutter_tools/lib/src/project.dart (hunk 2 above)
Comment (christopherfujino) [reply packages/flutter_tools/lib/src/project.dart:489]: Good find. Looks like whatever test you copy-pasted this from was also leaking the real file system. The whole purpose of the `testUsingContext()` function you're calling is that it allows you to provide `overrides` to the Zone/context getters (`globals.fs` is essentially calling `context.get<FileSystem>();).

Here is an example of constructing a FileSystem in `setUp` and then providing it as a context override for a `testUsingContext` invocation: https://github.com/flutter/flutter/blob/master/packages/flutter_tools/test/commands.shard/hermetic/build_test.dart#L60

Comment (christopherfujino) [reply packages/flutter_tools/lib/src/project.dart:489]: although if it makes sense, what you did here is probably better (it looks cleaner)

Comment (navaronbracke) [reply packages/flutter_tools/lib/src/project.dart:489]: Thanks for the tip, I'll update the test with an override for the file system.

File: packages/flutter_tools/test/general.shard/android/android_gradle_builder_test.dart [hunk 4]
```diff
@@ -2,10 +2,14 @@
 // Use of this source code is governed by a BSD-style license that can be
 // found in the LICENSE file.
 
+import 'dart:io';
```
Comment (christopherfujino) [inline packages/flutter_tools/test/general.shard/android/android_gradle_builder_test.dart:5]: avoid importing `dart:io`. most likely whatever definition you need is in either `flutter_tools/src/base/io.dart` or a similar library.

Comment (navaronbracke) [reply packages/flutter_tools/test/general.shard/android/android_gradle_builder_test.dart]: Done.

**Comment by navaronbracke** [comment]
@christopherfujino fixed the test (forgot the activity tag, woops) and now it no longer hangs when running the tests.
I wonder why it hung in the first place, perhaps the XML parser entering an infinite loop when it can't find an element?

**Review by christopherfujino — APPROVED** [review]
LGTM! Thanks for sticking through, I know this was tricky to test.

**Review by eliasyishak — APPROVED** [review]
👍

**Pull request merged**